common.sh

```diff
@@ -160,6 +163,243 @@ exit_fixed()
     exit $RESULT_FIXED
 }
 
+_get_cached_command()
+{
+    # Function gets a cached command
+    # input parameter is command provided by get_cached function
+    local command
+    command=$1
+    if [ ! -f "$command" ]; then
+        log_error "File $command does not exist. It is mandatory."
+    else
+        cat $command
+    fi
+}
+
+get_cached()
+{
+    # Get general system information from cache
+    # Usage:    get_cached file <passwd,group>
+    #           get_cached command <rpm_qa, rpm_Va, rpm_etc_Va, chconfig>
+    #           get_cached filelist <executable, allmyfiles, rpmrhsignedfiles, rpmtrackedfiles>
+    local command
+    local key
+    command=$1
+    key=$2
+    case $command in
+        "file")
+            case $key in
+                "passwd")
+                    ;;
+                "group")
+                    ;;
+                *)
+                    log_error "Unknown $key for $command. Supported are 'passwd', 'group'."
+                    return 1
+                    ;;
+            esac
+            _get_cached_command "$PREUPGRADE_CACHE/$key.log"
+            ;;
+        "command")
+            case $key in
+                "rpm_qa")
+                    ;;
+                "rpm_Va")
+                    ;;
+                "rpm_etc_Va")
+                    ;;
+                "chconfig")
+                    ;;
+                *)
+                    log_error "Unknown $key for $command. Supported are 'rpm_qa', 'rpm_Va', 'rpm_etc_Va', 'chkconfig'".
+                    return 1
+                    ;;
+            esac
+            _get_cached_command "$PREUPGRADE_CACHE/$key.log"
+            ;;
+        "filelist")
+            case $key in
+                "allmyfiles")
+                    ;;
+                "executable")
+                    ;;
+                "rpmtrackedfiles")
+                    ;;
+                "rpm_rhsigned")
+                    ;;
+                *)
+                    log_error "Unknown $key for $command. Supported are 'executable', 'allmyfiles', 'rpmtrackedfiles', 'rpm_rhsigned'."
+                    return 1
+            esac
+            _get_cached_command "$PREUPGRADE_CACHE/$key.log"
+            ;;
+        *)
+            log_error "Unknown command $command. Supported are 'file', 'command', filelist'."
+            return 1
+            ;;
+    esac
+
+}
+
+add_postupgrade()
+{
+    # Add postupgrade script to /root/preupgrade/postupgrade.d directory
+    # File as first parameter has to exists in modules directory
+    # Usage:    add_postupgrade file
+    local script_name
+    local postupgrade_name
+
+    script_name=$1
+    postupgrade_name="$PREUPG_POSTUPGRADE_DIR/$(basename $script_name)"
+    [ ! -f "$script_name" ] && log_error "$script_name does not exist." && exit_error
+    [ -f "$postupgrade_name" ] && log_warning "$script_name already exists in $PREUPG_POSTUPGRADE_DIR"
+    cp -af ${script_name} ${$postupgrade_name}
+}
```

Comment:
in addition, in case that script already exists, function shouldn't replace it by another, but should "return 1" or exit with error too - think about case that some postupgrade script resolves kernel panic and this function it will remove it. this should not happen.